File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/nodejs/signer.js
// Beckn protocol payload signing for DEG (Digital Energy Grid).
//
// Signs JSON payloads using Ed25519 + BLAKE2-512 and produces an Authorization
// header compatible with the Beckn protocol specification.
//
// Usage:
//
//   const signer = new PayloadSigner({
//     subscriberId:     'p2p-trading-sandbox1.com',
//     uniqueKeyId:      '76EU8aUqHouww7gawT6EibH4bseMCumyDv3sgyXSKENGk8NDcdVwmQ',
//     signingPrivateKey: 'Pc6dkYo5LeP0LkwvZXVRV9pcbeh8jDdtdHWymID5cjw=',
//   });
//
//   const authHeader = signer.signPayload(payloadBuffer);

'use strict';

const crypto = require('crypto');

// PKCS8 DER prefix for Ed25519 private key (wraps 32-byte seed).
const PKCS8_ED25519_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');

const DEFAULT_EXPIRY_SECONDS = 300; // 5 minutes

class PayloadSigner {
  /**
   * @param {object} config
   * @param {string} config.subscriberId      - Network participant ID (e.g. "p2p-trading-sandbox1.com")
   * @param {string} config.uniqueKeyId       - Unique key ID registered in the Beckn registry
   * @param {string} config.signingPrivateKey  - Base64-encoded Ed25519 seed (32 bytes)
   * @param {number} [config.expirySeconds=300] - Signature validity window in seconds (default 5 minutes)
   */
  constructor(config) {
    if (!config.subscriberId) {
      throw new Error('signer: subscriberId is required');
    }
    if (!config.uniqueKeyId) {
      throw new Error('signer: uniqueKeyId is required');
    }
    if (!config.signingPrivateKey) {
      throw new Error('signer: signingPrivateKey is required');
    }

    let seed;
    try {
      seed = Buffer.from(config.signingPrivateKey, 'base64');
    } catch (e) {
      throw new Error(`signer: invalid base64 private key: ${e.message}`);
    }

    // Validate that the base64 round-trips cleanly (rejects non-base64 input).
    if (seed.toString('base64') !== config.signingPrivateKey) {
      throw new Error('signer: invalid base64 private key: input is not valid base64');
    }

    if (seed.length !== 32) {
      throw new Error(`signer: private key seed must be 32 bytes, got ${seed.length}`);
    }

    this._subscriberId = config.subscriberId;
    this._uniqueKeyId = config.uniqueKeyId;
    this._privateKey = crypto.createPrivateKey({
      key: Buffer.concat([PKCS8_ED25519_PREFIX, seed]),
      format: 'der',
      type: 'pkcs8',
    });
    this._expirySeconds = config.expirySeconds ?? DEFAULT_EXPIRY_SECONDS;
  }

  /**
   * Sign a JSON payload and return the Authorization header value.
   * @param {Buffer|string} body - Raw JSON payload
   * @returns {string} Authorization header value
   */
  signPayload(body) {
    return this.signPayloadDetailed(body).authorizationHeader;
  }

  /**
   * Sign a JSON payload and return full signing details.
   * @param {Buffer|string} body - Raw JSON payload
   * @returns {{ authorizationHeader: string, createdAt: number, expiresAt: number, signature: string }}
   */
  signPayloadDetailed(body) {
    const now = Math.floor(Date.now() / 1000);
    return this._signPayloadAt(body, now);
  }

  /**
   * Sign at a specific Unix timestamp (used for deterministic testing).
   * @param {Buffer|string} body
   * @param {number} nowUnix - Unix timestamp in seconds
   * @returns {{ authorizationHeader: string, createdAt: number, expiresAt: number, signature: string }}
   */
  _signPayloadAt(body, nowUnix) {
    const createdAt = nowUnix;
    const expiresAt = nowUnix + this._expirySeconds;

    const bodyBuf = Buffer.isBuffer(body) ? body : Buffer.from(body);
    const signingString = buildSigningString(bodyBuf, createdAt, expiresAt);

    const sig = crypto.sign(null, Buffer.from(signingString), this._privateKey);
    const sigB64 = sig.toString('base64');

    const header =
      `Signature keyId="${this._subscriberId}|${this._uniqueKeyId}|ed25519"` +
      `,algorithm="ed25519"` +
      `,created="${createdAt}"` +
      `,expires="${expiresAt}"` +
      `,headers="(created) (expires) digest"` +
      `,signature="${sigB64}"`;

    return {
      authorizationHeader: header,
      createdAt,
      expiresAt,
      signature: sigB64,
    };
  }
}

/**
 * Build the canonical signing string:
 *   (created): {timestamp}
 *   (expires): {timestamp}
 *   digest: BLAKE-512={base64_hash}
 */
function buildSigningString(body, createdAt, expiresAt) {
  const hash = crypto.createHash('blake2b512');
  hash.update(body);
  const digest = hash.digest('base64');

  return `(created): ${createdAt}\n(expires): ${expiresAt}\ndigest: BLAKE-512=${digest}`;
}

module.exports = { PayloadSigner };

File: DEG-main/devkits/p2p-trading-ies-wave1/beckn-signing-kit/python/signer.py
"""Beckn protocol payload signing for DEG (Digital Energy Grid).

Signs JSON payloads using Ed25519 + BLAKE2-512 and produces an Authorization
header compatible with the Beckn protocol specification.

Usage::

    signer = PayloadSigner(
        subscriber_id='p2p-trading-sandbox1.com',
        unique_key_id='76EU8aUqHouww7gawT6EibH4bseMCumyDv3sgyXSKENGk8NDcdVwmQ',
        signing_private_key='Pc6dkYo5LeP0LkwvZXVRV9pcbeh8jDdtdHWymID5cjw=',
    )

    auth_header = signer.sign_payload(payload_bytes)
"""

from __future__ import annotations

import base64
import hashlib
import time
from dataclasses import dataclass

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

DEFAULT_EXPIRY_SECONDS = 300  # 5 minutes


@dataclass
class SignedResult:
    authorization_header: str
    created_at: int
    expires_at: int
    signature: str


class PayloadSigner:
    """Sign Beckn protocol payloads with Ed25519 + BLAKE2-512.

    Args:
        subscriber_id: Network participant ID (e.g. "p2p-trading-sandbox1.com").
        unique_key_id: Unique key ID registered in the Beckn registry.
        signing_private_key: Base64-encoded Ed25519 seed (32 bytes).
        expiry_seconds: Signature validity window in seconds (default 300).
    """

    def __init__(
        self,
        subscriber_id: str,
        unique_key_id: str,
        signing_private_key: str,
        expiry_seconds: int = DEFAULT_EXPIRY_SECONDS,
    ) -> None:
        if not subscriber_id:
            raise ValueError("signer: subscriber_id is required")
        if not unique_key_id:
            raise ValueError("signer: unique_key_id is required")
        if not signing_private_key:
            raise ValueError("signer: signing_private_key is required")

        try:
            seed = base64.b64decode(signing_private_key, validate=True)
        except Exception as exc:
            raise ValueError(f"signer: invalid base64 private key: {exc}") from exc

        # Validate that the base64 round-trips cleanly.
        if base64.b64encode(seed).decode() != signing_private_key:
            raise ValueError(
                "signer: invalid base64 private key: input is not valid base64"
            )

        if len(seed) != 32:
            raise ValueError(
                f"signer: private key seed must be 32 bytes, got {len(seed)}"
            )

        self._subscriber_id = subscriber_id
        self._unique_key_id = unique_key_id
        self._private_key = Ed25519PrivateKey.from_private_bytes(seed)
        self._expiry_seconds = expiry_seconds

    def sign_payload(self, body: bytes | str) -> str:
        """Sign a JSON payload and return the Authorization header value."""
        return self.sign_payload_detailed(body).authorization_header

    def sign_payload_detailed(self, body: bytes | str) -> SignedResult:
        """Sign a JSON payload and return full signing details."""
        now = int(time.time())
        return self._sign_payload_at(body, now)

    def _sign_payload_at(self, body: bytes | str, now_unix: int) -> SignedResult:
        """Sign at a specific Unix timestamp (used for deterministic testing)."""
        created_at = now_unix
        expires_at = now_unix + self._expiry_seconds

        body_bytes = body if isinstance(body, bytes) else body.encode("utf-8")
        signing_string = build_signing_string(body_bytes, created_at, expires_at)

        sig = self._private_key.sign(signing_string.encode("utf-8"))
        sig_b64 = base64.b64encode(sig).decode()

        header = (
            f'Signature keyId="{self._subscriber_id}|{self._unique_key_id}|ed25519"'
            f',algorithm="ed25519"'
            f',created="{created_at}"'
            f',expires="{expires_at}"'
            f',headers="(created) (expires) digest"'
            f',signature="{sig_b64}"'
        )

        return SignedResult(
            authorization_header=header,
            created_at=created_at,
            expires_at=expires_at,
            signature=sig_b64,
        )


def build_signing_string(body: bytes, created_at: int, expires_at: int) -> str:
    """Build the canonical signing string.

    Format::

        (created): {timestamp}
        (expires): {timestamp}
        digest: BLAKE-512={base64_hash}
    """
    digest = base64.b64encode(
        hashlib.blake2b(body, digest_size=64).digest()
    ).decode()

    return f"(created): {created_at}\n(expires): {expires_at}\ndigest: BLAKE-512={digest}"
